posts/draft2.md

```diff
@@ -0,0 +1,252 @@
+
+```python
+    def parse(self, inputstring: str, document: nodes.document) -> None:
+        """Parse source text.
+        :param inputstring: The source string to parse
+        :param document: The root docutils node to add AST elements to
+        """
+        config = document.settings.env.myst_config
+        parser = default_parser(config)
+        parser.options["document"] = document
+        env = AttrDict()
+```
+
+to parse the input text into a token stream and then rendering those (via the
+SphinxRenderer which is a subclass of the DocutilsRenderer):
+
+```python
+        tokens = parser.parse(inputstring, env)
+        if not tokens or tokens[0].type != "front_matter":
+            # we always add front matter, so that we can merge it with global keys,
+            # specified in the sphinx configuration
+            tokens = [Token("front_matter", "", 0, content="{}", map=[0, 0])] + tokens
+        parser.renderer.render(tokens, parser.options, env)
+```
+
+into a Docutils' document representation: 
+
+```python
+def parse(app: Sphinx, text: str, docname: str = "index") -> nodes.document:
+    """Parse a string as MystMarkdown with Sphinx application."""
+    app.env.temp_data["docname"] = docname
+    app.env.all_docs[docname] = time.time()
+    reader = SphinxStandaloneReader()
+    reader.setup(app)
+    parser = MystParser()
+    parser.set_application(app)
+    with sphinx_domains(app.env):
+        return publish_doctree(
+            text,
+            path.join(app.srcdir, docname + ".md"),
+            reader=reader,
+            parser=parser,
+            parser_name="markdown",
+            settings_overrides={"env": app.env, "gettext_compact": True},
+        )
+```
+
+Finally, those objects are passed to the built-in Sphinx machinery to write the html output.
+
+## The big question
+
+This is great! We now understand why we need Sphinx to generate the expected HTML output
+for roles and directives. Several questions now arise:
+
+* Do we want to have Sphinx-free support for rendering roles and directives in the MyST-Parser 
+
+    Nikola (among other projects) already have their own machinery (based in Docutils) to
+    build the final HTML output. Getting the docutils object from the Python API would be
+    a super interesting way to easily expose and provide that object to the downstream
+    projects!
+
+    One caveat with this approach would be missing some Sphinx features (ie. cross-linking
+    between documents) based on custom roles and directives that we may need to re-implement.
+
+2. Do we want to have docutils-free support for roles and directives in the MyST Python API?
+
+    Docutils actually introduces the roles and directives concept (that Sphinx extend) so if
+    we want to go docutils-free, then we will need to re-implement those concepts.
+
+3. Does it makes sense to create a docutils alternative in Python? At what extend?
```

Comment:
```suggestion
3. Does it makes sense to create a docutils alternative in Python? How much of its functionality would need to be replicated? How should it be extended or enhanced?
```

Reply:
how about that?

(also, very intriguing question that I am a little bit terrified of, but is a very important question!)

Reply:
I like that proposal, and it captures the spirit of what I was asking for with the "At what extend?"